DOCSP-41306: schema version trait (#3051)
jenssegers/laravel-mongodb: docs/eloquent-models/model-class.txt

Create a Versioned Model Schema
-------------------------------

You can implement a schema versioning pattern into your application by
using the ``HasSchemaVersion`` trait on an Eloquent model. You might
choose to implement a schema version to organize or standardize a
collection that contains data with different schemas.

.. tip::

   To learn more about schema versioning, see the :manual:`Model Data for
   Schema Versioning </tutorial/model-data-for-schema-versioning/>`
   tutorial in the {+server-docs-name+}.

To use this feature with models that use MongoDB as a database, add the
``MongoDB\Laravel\Eloquent\HasSchemaVersion`` import to your model.
Then, set the ``SCHEMA_VERSION`` constant to ``1`` to set the first
schema version on your collection. If your collection evolves to contain
multiple schemas, you can update the value of the ``SCHEMA_VERSION``
constant in subsequent model classes.

When creating your model, you can define the ``migrateSchema()`` method
to specify a migration to the current schema version upon retrieving a
model. In this method, you can specify the changes to make to an older
model to update it to match the current schema version.

> When you save a model that does not have a schema version
> specified, Laravel assumes that it follows the latest schema version.
> When you retrieve a model that does not contain the ``schema_version``
> field, Laravel assumes that its schema version is ``0`` and performs the
> migration.

Reviewer: ```suggestion
When you save a model that does not have a schema version
specified, the trait assumes that it follows the latest schema version.
When you retrieve a model that does not contain the ``schema_version``
field, the trait assumes that its schema version is ``0`` and performs the
migration.
```

   :emphasize-lines: 5,10,12
   :dedent:

.. _laravel-schema-versioning:

Create a Versioned Model Schema
-------------------------------

You can implement a schema versioning pattern into your application by
using the ``HasSchemaVersion`` trait on an Eloquent model. You might
choose to implement a schema version to organize or standardize a
collection that contains data with different schemas.

.. tip::

   To learn more about schema versioning, see the :manual:`Model Data for
   Schema Versioning </tutorial/model-data-for-schema-versioning/>`
   tutorial in the {+server-docs-name+}.

To use this feature with models that use MongoDB as a database, add the
``MongoDB\Laravel\Eloquent\HasSchemaVersion`` import to your model.
Then, set the ``SCHEMA_VERSION`` constant to set the current schema

> version for your collection.

Reviewer: Should be 1 for the first version of the data doesn't contain any versioned schema.

When creating your model, you can define the ``migrateSchema()`` method
to specify a migration to the current schema version upon saving a
model.

Schema Versioning Example
~~~~~~~~~~~~~~~~~~~~~~~~~

The example class is defined with the following behavior:

- Implements the ``HasSchemaVersion`` trait and sets the current
  ``SCHEMA_VERSION`` to ``2``

- Defines the ``migrateSchema()`` method to migrate models in which the
  schema version is less than ``2`` to have a ``galaxy`` field that has a value
  of ``'Milky Way'``

.. literalinclude:: /includes/eloquent-models/PlanetSchemaVersion.php
   :language: php
   :emphasize-lines: 10,12,14
   :dedent:

When you save a model in which the ``schema_version`` field value is
less than ``2``, Laravel adds the ``galaxy`` field and updates the schema
version to the current version (``2``).

The following code creates and saves instances of the ``Planet`` model that are

> not at the current schema version, then retrieves the planets from the

Reviewer 2: ```suggestion
not at the current schema version, then retrieves the Planet documents[? objects?] from the
```

collection to demonstrate the migration changes:

.. io-code-block::
   :copyable: true

   .. input::
      :language: php

      $saturn = Planet::create([
          'name' => 'Saturn',
          'type' => 'gas',
      ]);

      $wasp = Planet::create([
          'name' => 'WASP-39 b',
          'type' => 'gas',
          'schema_version' => 1,
      ]);

      $planets = Planet::where('type', 'gas')

>           ->get();

Reviewer: Code snippets can be extracted from the test file like this: https://github.com/mongodb/laravel-mongodb/blob/a62d4b91796b00c2cf46845bab9fd23cc0df6a97/docs/includes/query-builder/QueryBuilderTest.php#L51-L56

   .. output::
      :language: none
      :visible: false

      [
        {
          "_id": ...,
          "name": "Saturn",
          "type": "gas",

>           "galaxy": "Milky Way",

Reviewer: We assume a document inserted without specifying the version uses the latest schema version. So it will not have this field by the migration. 
You should write a unit test to ensure your example is correct.

   :emphasize-lines: 5,10,12
   :dedent:

.. _laravel-schema-versioning:

Create a Versioned Model Schema
-------------------------------

You can implement a schema versioning pattern into your application by
using the ``HasSchemaVersion`` trait on an Eloquent model. You might
choose to implement a schema version to organize or standardize a
collection that contains data with different schemas.

.. tip::

   To learn more about schema versioning, see the :manual:`Model Data for
   Schema Versioning </tutorial/model-data-for-schema-versioning/>`
   tutorial in the {+server-docs-name+}.

To use this feature with models that use MongoDB as a database, add the
``MongoDB\Laravel\Eloquent\HasSchemaVersion`` import to your model.
Then, set the ``SCHEMA_VERSION`` constant to ``1`` to set the first
schema version on your collection. If your collection evolves to contain
multiple schemas, you can update the value of the ``SCHEMA_VERSION``
constant in subsequent model classes.

> When creating your model, you can define the ``migrateSchema()`` method
> to specify a migration to the current schema version upon saving a
> model. In this method, you can specify the changes to make to an older
> model to update it to match the current schema version.

Reviewer: ```suggestion
When creating your model, you can define the ``migrateSchema()`` method
to specify a migration to the current schema version upon retrieving a
model. In this method, you can specify the changes to make to an older
model to update it to match the current schema version.
```

Schema Versioning Example
~~~~~~~~~~~~~~~~~~~~~~~~~

In this sample situation, you are working with a collection that was
first modeled by the following class:

.. literalinclude:: /includes/eloquent-models/PlanetSchemaVersion1.php
   :language: php
   :dedent:

Now, you want to implement a new schema version on the collection.
You can define the new model class with the following behavior:

- Implements the ``HasSchemaVersion`` trait and sets the current
  ``SCHEMA_VERSION`` to ``2``

- Defines the ``migrateSchema()`` method to migrate models in which the
  schema version is less than ``2`` to have a ``galaxy`` field that has a value
  of ``'Milky Way'``

.. literalinclude:: /includes/eloquent-models/PlanetSchemaVersion2.php
   :language: php
   :emphasize-lines: 10,12,20
   :dedent:

> When you save or retrieve a model that does not have a schema version
> specified, Laravel assumes that it follows the latest schema version.

Reviewer: We automatically save the current "SCHEMA_VERSION". When a model is retrieved, if it doesn't contain the schema_version property, we assume it was inserted before the model was versioned, so we assume its version `0` and run the full migration.

```suggestion
When you save a model that does not have a schema version specified,
Laravel assumes that it follows the latest schema version.
```

Author: Will update the copy
